File: macos-helper/Tests/CaptureCoreTests/CaptureAPIClientTests.swift
import Foundation
import Testing
@testable import CaptureCore

// MARK: - In-process mock transport

final class MockURLProtocol: URLProtocol {
    struct Stub {
        let status: Int
        let body: Data
        var headers: [String: String] = [:]
    }

    struct RecordedRequest {
        let method: String
        let url: URL
        let headers: [String: String]
        let body: Data?
    }

    nonisolated(unsafe) static var handler: ((RecordedRequest) -> Stub)?
    nonisolated(unsafe) private static var recordedStorage: [RecordedRequest] = []
    private static let lock = NSLock()

    static var recorded: [RecordedRequest] {
        lock.lock(); defer { lock.unlock() }
        return recordedStorage
    }

    static func reset() {
        lock.lock()
        handler = nil
        recordedStorage = []
        lock.unlock()
    }

    private static func record(_ request: RecordedRequest) -> Stub {
        lock.lock(); defer { lock.unlock() }
        recordedStorage.append(request)
        guard let handler else {
            return Stub(status: 500, body: Data("no handler installed".utf8))
        }
        return handler(request)
    }

    override class func canInit(with request: URLRequest) -> Bool { true }
    override class func canonicalRequest(for request: URLRequest) -> URLRequest { request }
    override func stopLoading() {}

    override func startLoading() {
        let recordedRequest = MockURLProtocol.RecordedRequest(
            method: request.httpMethod ?? "GET",
            url: request.url!,
            headers: request.allHTTPHeaderFields ?? [:],
            body: Self.bodyData(of: request)
        )
        let stub = Self.record(recordedRequest)
        let response = HTTPURLResponse(
            url: request.url!,
            statusCode: stub.status,
            httpVersion: "HTTP/1.1",
            headerFields: stub.headers.merging(["Content-Type": "application/json"]) { a, _ in a }
        )!
        client?.urlProtocol(self, didReceive: response, cacheStoragePolicy: .notAllowed)
        client?.urlProtocol(self, didLoad: stub.body)
        client?.urlProtocolDidFinishLoading(self)
    }

    /// URLProtocol never sees `httpBody` — only `httpBodyStream`.
    private static func bodyData(of request: URLRequest) -> Data? {
        if let body = request.httpBody { return body }
        guard let stream = request.httpBodyStream else { return nil }
        stream.open()
        defer { stream.close() }
        var data = Data()
        let bufferSize = 64 * 1024
        var buffer = [UInt8](repeating: 0, count: bufferSize)
        while stream.hasBytesAvailable {
            let read = stream.read(&buffer, maxLength: bufferSize)
            guard read > 0 else { break }
            data.append(buffer, count: read)
        }
        return data
    }
}

// MARK: - Tests
//
// .serialized: the mock transport uses process-global state, so these tests
// must not run in parallel with each other.

@Suite(.serialized) final class CaptureAPIClientTests {

    private let client: CaptureAPIClient
    private let tempDir: URL

    init() throws {
        MockURLProtocol.reset()
        let configuration = URLSessionConfiguration.ephemeral
        configuration.protocolClasses = [MockURLProtocol.self]
        let session = URLSession(configuration: configuration)
        client = CaptureAPIClient(
            baseURL: URL(string: "https://crm.example.com")!,
            token: "agbcap_deadbeef",
            urlSession: session
        )
        tempDir = FileManager.default.temporaryDirectory
            .appendingPathComponent("agb-client-tests-\(UUID().uuidString)", isDirectory: true)
        try FileManager.default.createDirectory(at: tempDir, withIntermediateDirectories: true)
    }

    deinit {
        MockURLProtocol.reset()
        try? FileManager.default.removeItem(at: tempDir)
    }

    private func json(_ string: String) -> Data { Data(string.utf8) }

    private func writeChunkFile(seq: Int, pcmBytes: Int = 640) throws -> URL {
        let url = tempDir.appendingPathComponent(ChunkSpooler.chunkFileName(seq: seq))
        try WavCodec.wrap(pcm: Data(repeating: UInt8(seq + 1), count: pcmBytes)).write(to: url)
        return url
    }

    // MARK: ping

    @Test func pingSendsAuthAndProtocolHeaders() async throws {
        MockURLProtocol.handler = { request in
            .init(status: 200,
                  body: Data(#"{"ok":true,"workspaceId":"ws1","userId":"u1","retentionDays":30}"#.utf8))
        }
        let pong = try await client.ping()
        #expect(pong.ok)
        #expect(pong.workspaceId == "ws1")
        #expect(pong.userId == "u1")
        #expect(pong.retentionDays == 30)

        let request = try #require(MockURLProtocol.recorded.first)
        #expect(request.method == "GET")
        #expect(request.url.path == "/api/capture/ping")
        #expect(request.headers["Authorization"] == "Bearer agbcap_deadbeef")
        #expect(request.headers["X-Capture-Protocol"] == "1")
    }

    @Test func ping401MapsToUnauthorized() async {
        MockURLProtocol.handler = { _ in .init(status: 401, body: Data()) }
        await #expect(throws: CaptureAPIClient.APIError.self) {
            _ = try await self.client.ping()
        }
        do {
            _ = try await client.ping()
        } catch let error as CaptureAPIClient.APIError {
            guard case .unauthorized = error else {
                Issue.record("expected .unauthorized, got \(error)")
                return
            }
        } catch {
            Issue.record("unexpected error type \(error)")
        }
    }

    // MARK: createSession

    @Test func createSessionBodyMatchesProtocol() async throws {
        MockURLProtocol.handler = { _ in
            .init(status: 201, body: Data(#"{"sessionId":"sess-42"}"#.utf8))
        }
        let sessionId = try await client.createSession(
            meta: .init(startedAt: Date(), sourceApp: "WhatsApp"))
        #expect(sessionId == "sess-42")

        let request = try #require(MockURLProtocol.recorded.first)
        #expect(request.method == "POST")
        #expect(request.url.path == "/api/capture/sessions")
        let rawBody = try #require(request.body)
        let body = try #require(try JSONSerialization.jsonObject(with: rawBody) as? [String: Any])
        #expect(body["sampleRate"] as? Int == 16_000)
        #expect(body["channels"] as? Int == 2)
        #expect(body["format"] as? String == "wav-pcm16")
        #expect(body["helperVersion"] as? String == AudioConstants.helperVersion)
        #expect(body["sourceApp"] as? String == "WhatsApp")
        #expect(body["startedAt"] is String)
    }

    @Test func createSessionEncodesNullSourceApp() async throws {
        MockURLProtocol.handler = { _ in
            .init(status: 201, body: Data(#"{"sessionId":"s"}"#.utf8))
        }
        _ = try await client.createSession(meta: .init(startedAt: Date(), sourceApp: nil))
        let request = try #require(MockURLProtocol.recorded.first)
        let rawBody = try #require(request.body)
        let body = try #require(try JSONSerialization.jsonObject(with: rawBody) as? [String: Any])
        #expect(body["sourceApp"] is NSNull, "protocol requires explicit null")
    }

    // MARK: uploadChunk

    @Test func uploadChunkPutsWavBodyWithContentType() async throws {
        let fileURL = try writeChunkFile(seq: 3, pcmBytes: 1_000)
        let expectedBody = try Data(contentsOf: fileURL)

        MockURLProtocol.handler = { _ in
            .init(status: 200, body: Data(#"{"ok":true,"bytes":1044}"#.utf8))
        }
        let bytes = try await client.uploadChunk(sessionId: "sess-1", seq: 3, fileURL: fileURL)
        #expect(bytes == 1_044)

        let request = try #require(MockURLProtocol.recorded.first)
        #expect(request.method == "PUT")
        #expect(request.url.path == "/api/capture/sessions/sess-1/chunks/3")
        #expect(request.headers["Content-Type"] == "audio/wav")
        #expect(request.headers["Authorization"] == "Bearer agbcap_deadbeef")
        #expect(request.body == expectedBody, "raw body must be the exact WAV bytes")
    }

    @Test func uploadChunk413MapsToPayloadTooLarge() async throws {
        let fileURL = try writeChunkFile(seq: 0)
        MockURLProtocol.handler = { _ in .init(status: 413, body: Data()) }
        do {
            _ = try await client.uploadChunk(sessionId: "s", seq: 0, fileURL: fileURL)
            Issue.record("expected payloadTooLarge")
        } catch let error as CaptureAPIClient.APIError {
            guard case .payloadTooLarge = error else {
                Issue.record("expected .payloadTooLarge, got \(error)")
                return
            }
        }
    }

    // MARK: finalize + 409 recovery

    @Test func finalize409ThrowsMissingChunks() async throws {
        MockURLProtocol.handler = { _ in
            .init(status: 409, body: Data(#"{"missing":[1,3]}"#.utf8))
        }
        do {
            _ = try await client.finalize(
                sessionId: "s",
                body: .init(endedAt: Date(), durationSecs: 90, totalChunks: 4, partial: false))
            Issue.record("expected missingChunks")
        } catch let error as CaptureAPIClient.APIError {
            guard case .missingChunks(let seqs) = error else {
                Issue.record("expected .missingChunks, got \(error)")
                return
            }
            #expect(seqs == [1, 3])
        }
    }

    @Test func finalizeRecoveringReuploadsMissingThenRetries() async throws {
        let chunk1 = try writeChunkFile(seq: 1)
        let lock = NSLock()
        var finalizeCalls = 0

        MockURLProtocol.handler = { request in
            if request.url.path.hasSuffix("/finalize") {
                lock.lock(); finalizeCalls += 1; let n = finalizeCalls; lock.unlock()
                if n == 1 {
                    return .init(status: 409, body: Data(#"{"missing":[1]}"#.utf8))
                }
                return .init(
                    status: 200,
                    body: Data(#"{"ok":true,"recordingId":"rec-9","title":"Test call","brief":"b","actionItemCount":2,"contact":null,"suspectFlags":[]}"#.utf8))
            }
            return .init(status: 200, body: Data(#"{"ok":true,"bytes":1}"#.utf8))
        }

        let result = try await client.finalizeRecovering(
            sessionId: "sess-1",
            body: .init(endedAt: Date(), durationSecs: 60, totalChunks: 2, partial: false)
        ) { seq in
            seq == 1 ? chunk1 : nil
        }

        #expect(result.ok)
        #expect(result.recordingId == "rec-9")
        #expect(result.title == "Test call")
        #expect(result.actionItemCount == 2)
        #expect(!result.raw.isEmpty)

        let paths = MockURLProtocol.recorded.map { "\($0.method) \($0.url.path)" }
        #expect(paths == [
            "POST /api/capture/sessions/sess-1/finalize",
            "PUT /api/capture/sessions/sess-1/chunks/1",
            "POST /api/capture/sessions/sess-1/finalize",
        ], "409 → re-upload missing seq → retry finalize")
    }

    @Test func finalizeRecoveringGivesUpWhenChunkFileMissing() async throws {
        MockURLProtocol.handler = { request in
            if request.url.path.hasSuffix("/finalize") {
                return .init(status: 409, body: Data(#"{"missing":[7]}"#.utf8))
            }
            return .init(status: 200, body: Data())
        }
        do {
            _ = try await client.finalizeRecovering(
                sessionId: "s",
                body: .init(endedAt: Date(), durationSecs: 10, totalChunks: 8, partial: false)
            ) { _ in nil }
            Issue.record("expected missingChunks")
        } catch let error as CaptureAPIClient.APIError {
            guard case .missingChunks(let seqs) = error else {
                Issue.record("expected .missingChunks, got \(error)")
                return
            }
            #expect(seqs == [7])
        }
    }

    // MARK: abandon

    @Test func abandonSendsDelete() async throws {
        MockURLProtocol.handler = { _ in
            .init(status: 200, body: Data(#"{"ok":true}"#.utf8))
        }
        try await client.abandon(sessionId: "sess-1")
        let request = try #require(MockURLProtocol.recorded.first)
        #expect(request.method == "DELETE")
        #expect(request.url.path == "/api/capture/sessions/sess-1")
    }

    // MARK: backoff

    @Test func exponentialBackoffSequenceCappedAt60() {
        var backoff = ExponentialBackoff()
        var delays: [TimeInterval] = []
        for _ in 0..<9 { delays.append(backoff.nextDelay()) }
        #expect(delays == [1, 2, 4, 8, 16, 32, 60, 60, 60])

        backoff.reset()
        #expect(backoff.nextDelay() == 1)
    }
}

File: windows-helper/CaptureCore.Tests/CaptureApiClientTests.cs
using System.Net;
using System.Text;
using System.Text.Json;
using AGB.CaptureCore;
using Xunit;

namespace AGB.CaptureCore.Tests;

/// <summary>
/// Port of <c>CaptureAPIClientTests.swift</c>. The Swift suite injected a custom
/// URLSession with a <c>MockURLProtocol</c>; here we inject a
/// <see cref="MockHttpMessageHandler"/> into the client — the same testing seam.
/// </summary>
public class CaptureApiClientTests : IDisposable
{
    private sealed record RecordedRequest(string Method, Uri Url, IReadOnlyDictionary<string, string> Headers, byte[]? Body);

    private sealed class Stub
    {
        public int Status { get; init; } = 200;
        public byte[] Body { get; init; } = Array.Empty<byte>();
    }

    private sealed class MockHttpMessageHandler : HttpMessageHandler
    {
        public Func<RecordedRequest, Stub>? Handler { get; set; }
        public List<RecordedRequest> Recorded { get; } = new();

        protected override async Task<HttpResponseMessage> SendAsync(HttpRequestMessage request, CancellationToken ct)
        {
            byte[]? body = request.Content is null ? null : await request.Content.ReadAsByteArrayAsync(ct);
            var headers = new Dictionary<string, string>(StringComparer.OrdinalIgnoreCase);
            foreach (var h in request.Headers)
                headers[h.Key] = string.Join(",", h.Value);
            if (request.Content is not null)
                foreach (var h in request.Content.Headers)
                    headers[h.Key] = string.Join(",", h.Value);

            var recorded = new RecordedRequest(request.Method.Method, request.RequestUri!, headers, body);
            Recorded.Add(recorded);

            Stub stub = Handler?.Invoke(recorded) ?? new Stub { Status = 500, Body = "no handler installed"u8.ToArray() };
            return new HttpResponseMessage((HttpStatusCode)stub.Status)
            {
                Content = new ByteArrayContent(stub.Body),
            };
        }
    }

    private readonly MockHttpMessageHandler _handler = new();
    private readonly CaptureApiClient _client;
    private readonly string _tempDir;

    public CaptureApiClientTests()
    {
        _client = new CaptureApiClient(
            new Uri("https://crm.example.com"),
            "agbcap_deadbeef",
            _handler);
        _tempDir = Path.Combine(Path.GetTempPath(), $"agb-client-tests-{Guid.NewGuid()}");
        Directory.CreateDirectory(_tempDir);
    }

    public void Dispose()
    {
        try { Directory.Delete(_tempDir, recursive: true); } catch { /* ignore */ }
    }

    private string WriteChunkFile(int seq, int pcmBytes = 640)
    {
        string path = Path.Combine(_tempDir, ChunkSpooler.ChunkFileName(seq));
        var pcm = new byte[pcmBytes];
        Array.Fill(pcm, (byte)(seq + 1));
        File.WriteAllBytes(path, WavCodec.Wrap(pcm));
        return path;
    }

    // ------------------------------------------------------------------- ping

    [Fact]
    public async Task PingSendsAuthAndProtocolHeaders()
    {
        _handler.Handler = _ => new Stub
        {
            Status = 200,
            Body = """{"ok":true,"workspaceId":"ws1","userId":"u1","retentionDays":30}"""u8.ToArray(),
        };
        var pong = await _client.PingAsync();
        Assert.True(pong.Ok);
        Assert.Equal("ws1", pong.WorkspaceId);
        Assert.Equal("u1", pong.UserId);
        Assert.Equal(30, pong.RetentionDays);

        var request = _handler.Recorded[0];
        Assert.Equal("GET", request.Method);
        Assert.Equal("/api/capture/ping", request.Url.AbsolutePath);
        Assert.Equal("Bearer agbcap_deadbeef", request.Headers["Authorization"]);
        Assert.Equal("1", request.Headers["X-Capture-Protocol"]);
    }

    [Fact]
    public async Task Ping401MapsToUnauthorized()
    {
        _handler.Handler = _ => new Stub { Status = 401 };
        var ex = await Assert.ThrowsAsync<CaptureApiException>(() => _client.PingAsync());
        Assert.Equal(CaptureApiErrorKind.Unauthorized, ex.Kind);
    }

    // -------------------------------------------------------------- createSession

    [Fact]
    public async Task CreateSessionBodyMatchesProtocol()
    {
        _handler.Handler = _ => new Stub { Status = 201, Body = """{"sessionId":"sess-42"}"""u8.ToArray() };
        string sessionId = await _client.CreateSessionAsync(new SessionMeta(DateTimeOffset.UtcNow, "WhatsApp"));
        Assert.Equal("sess-42", sessionId);

        var request = _handler.Recorded[0];
        Assert.Equal("POST", request.Method);
        Assert.Equal("/api/capture/sessions", request.Url.AbsolutePath);

        using var doc = JsonDocument.Parse(request.Body!);
        var root = doc.RootElement;
        Assert.Equal(16_000, root.GetProperty("sampleRate").GetInt32());
        Assert.Equal(2, root.GetProperty("channels").GetInt32());
        Assert.Equal("wav-pcm16", root.GetProperty("format").GetString());
        Assert.Equal(AudioConstants.HelperVersion, root.GetProperty("helperVersion").GetString());
        Assert.Equal("WhatsApp", root.GetProperty("sourceApp").GetString());
        Assert.Equal(JsonValueKind.String, root.GetProperty("startedAt").ValueKind);
    }

    [Fact]
    public async Task CreateSessionEncodesNullSourceApp()
    {
        _handler.Handler = _ => new Stub { Status = 201, Body = """{"sessionId":"s"}"""u8.ToArray() };
        await _client.CreateSessionAsync(new SessionMeta(DateTimeOffset.UtcNow, null));

        using var doc = JsonDocument.Parse(_handler.Recorded[0].Body!);
        Assert.Equal(JsonValueKind.Null, doc.RootElement.GetProperty("sourceApp").ValueKind);
    }

    // ----------------------------------------------------------------- uploadChunk

    [Fact]
    public async Task UploadChunkPutsWavBodyWithContentType()
    {
        string filePath = WriteChunkFile(seq: 3, pcmBytes: 1_000);
        byte[] expectedBody = File.ReadAllBytes(filePath);

        _handler.Handler = _ => new Stub { Status = 200, Body = """{"ok":true,"bytes":1044}"""u8.ToArray() };
        int bytes = await _client.UploadChunkAsync("sess-1", 3, filePath);
        Assert.Equal(1_044, bytes);

        var request = _handler.Recorded[0];
        Assert.Equal("PUT", request.Method);
        Assert.Equal("/api/capture/sessions/sess-1/chunks/3", request.Url.AbsolutePath);
        Assert.Equal("audio/wav", request.Headers["Content-Type"]);
        Assert.Equal("Bearer agbcap_deadbeef", request.Headers["Authorization"]);
        Assert.Equal(expectedBody, request.Body); // raw body must be the exact WAV bytes
    }

    [Fact]
    public async Task UploadChunk413MapsToPayloadTooLarge()
    {
        string filePath = WriteChunkFile(seq: 0);
        _handler.Handler = _ => new Stub { Status = 413 };
        var ex = await Assert.ThrowsAsync<CaptureApiException>(
            () => _client.UploadChunkAsync("s", 0, filePath));
        Assert.Equal(CaptureApiErrorKind.PayloadTooLarge, ex.Kind);
    }

    // ----------------------------------------------------- finalize + 409 recovery

    [Fact]
    public async Task Finalize409ThrowsMissingChunks()
    {
        _handler.Handler = _ => new Stub { Status = 409, Body = """{"missing":[1,3]}"""u8.ToArray() };
        var ex = await Assert.ThrowsAsync<CaptureApiException>(
            () => _client.FinalizeAsync("s", new FinalizeBody(DateTimeOffset.UtcNow, 90, 4, false)));
        Assert.Equal(CaptureApiErrorKind.MissingChunks, ex.Kind);
        Assert.Equal(new[] { 1, 3 }, ex.MissingSeqs);
    }

    [Fact]
    public async Task FinalizeRecoveringReuploadsMissingThenRetries()
    {
        string chunk1 = WriteChunkFile(seq: 1);
        int finalizeCalls = 0;
        var gate = new object();

        _handler.Handler = request =>
        {
            if (request.Url.AbsolutePath.EndsWith("/finalize", StringComparison.Ordinal))
            {
                int n;
                lock (gate) { n = ++finalizeCalls; }
                if (n == 1)
                    return new Stub { Status = 409, Body = """{"missing":[1]}"""u8.ToArray() };
                return new Stub
                {
                    Status = 200,
                    Body = """{"ok":true,"recordingId":"rec-9","title":"Test call","brief":"b","actionItemCount":2,"contact":null,"suspectFlags":[]}"""u8.ToArray(),
                };
            }
            return new Stub { Status = 200, Body = """{"ok":true,"bytes":1}"""u8.ToArray() };
        };

        var result = await _client.FinalizeRecoveringAsync(
            "sess-1",
            new FinalizeBody(DateTimeOffset.UtcNow, 60, 2, false),
            chunkFilePath: seq => seq == 1 ? chunk1 : null);

        Assert.True(result.Ok);
        Assert.Equal("rec-9", result.RecordingId);
        Assert.Equal("Test call", result.Title);
        Assert.Equal(2, result.ActionItemCount);
        Assert.NotEmpty(result.Raw);

        var paths = _handler.Recorded.Select(r => $"{r.Method} {r.Url.AbsolutePath}").ToArray();
        Assert.Equal(new[]
        {
            "POST /api/capture/sessions/sess-1/finalize",
            "PUT /api/capture/sessions/sess-1/chunks/1",
            "POST /api/capture/sessions/sess-1/finalize",
        }, paths); // 409 → re-upload missing seq → retry finalize
    }

    [Fact]
    public async Task FinalizeRecoveringGivesUpWhenChunkFileMissing()
    {
        _handler.Handler = request =>
            request.Url.AbsolutePath.EndsWith("/finalize", StringComparison.Ordinal)
                ? new Stub { Status = 409, Body = """{"missing":[7]}"""u8.ToArray() }
                : new Stub { Status = 200 };

        var ex = await Assert.ThrowsAsync<CaptureApiException>(
            () => _client.FinalizeRecoveringAsync(
                "s",
                new FinalizeBody(DateTimeOffset.UtcNow, 10, 8, false),
                chunkFilePath: _ => null));
        Assert.Equal(CaptureApiErrorKind.MissingChunks, ex.Kind);
        Assert.Equal(new[] { 7 }, ex.MissingSeqs);
    }

    // ----------------------------------------------------------------- abandon

    [Fact]
    public async Task AbandonSendsDelete()
    {
        _handler.Handler = _ => new Stub { Status = 200, Body = """{"ok":true}"""u8.ToArray() };
        await _client.AbandonAsync("sess-1");
        var request = _handler.Recorded[0];
        Assert.Equal("DELETE", request.Method);
        Assert.Equal("/api/capture/sessions/sess-1", request.Url.AbsolutePath);
    }

    // ----------------------------------------------------------------- backoff

    [Fact]
    public void ExponentialBackoffSequenceCappedAt60()
    {
        var backoff = new ExponentialBackoff();
        var delays = new List<double>();
        for (int i = 0; i < 9; i++) delays.Add(backoff.NextDelay().TotalSeconds);
        Assert.Equal(new double[] { 1, 2, 4, 8, 16, 32, 60, 60, 60 }, delays);

        backoff.Reset();
        Assert.Equal(1, backoff.NextDelay().TotalSeconds);
    }
}
